File: backend/app/services/voice.py
"""ElevenLabs voice integration: STT and TTS."""

import logging
import httpx
from ..config import settings

logger = logging.getLogger(__name__)

ELEVENLABS_BASE = "https://api.elevenlabs.io/v1"
DEFAULT_VOICE_ID = "21m00Tcm4TlvDq8ikWAM"  # Rachel


async def transcribe_audio(audio_bytes: bytes) -> str:
    """Convert audio to text via ElevenLabs STT."""
    async with httpx.AsyncClient(timeout=30.0) as client:
        resp = await client.post(
            f"{ELEVENLABS_BASE}/speech-to-text",
            headers={"xi-api-key": settings.ELEVENLABS_API_KEY},
            files={"file": ("recording.webm", audio_bytes, "audio/webm")},
            data={"model_id": "scribe_v1"},
        )
        resp.raise_for_status()
        return resp.json()["text"]


async def text_to_speech(text: str) -> bytes:
    """Convert text to audio via ElevenLabs TTS."""
    async with httpx.AsyncClient(timeout=30.0) as client:
        resp = await client.post(
            f"{ELEVENLABS_BASE}/text-to-speech/{DEFAULT_VOICE_ID}",
            headers={
                "xi-api-key": settings.ELEVENLABS_API_KEY,
                "Content-Type": "application/json",
            },
            json={
                "text": text,
                "model_id": "eleven_multilingual_v2",
            },
            params={"output_format": "mp3_44100_128"},
        )
        resp.raise_for_status()
        return resp.content

File: frontend/src/api/voice.ts
const BASE_URL = "/api";

export async function transcribeAudio(audio: Blob): Promise<string> {
  const formData = new FormData();
  formData.append("audio", audio, "recording.webm");

  const resp = await fetch(`${BASE_URL}/voice/transcribe`, {
    method: "POST",
    body: formData,
  });

  if (!resp.ok) {
    const err = await resp.json().catch(() => ({ detail: "Transcription failed" }));
    throw new Error(err.detail || `HTTP ${resp.status}`);
  }

  const data = await resp.json();
  return data.text;
}

export async function textToSpeech(text: string): Promise<Blob> {
  const resp = await fetch(`${BASE_URL}/voice/speak`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ text }),
  });

  if (!resp.ok) {
    const err = await resp.json().catch(() => ({ detail: "TTS failed" }));
    throw new Error(err.detail || `HTTP ${resp.status}`);
  }

  return resp.blob();
}
